File: src/language/service/symbols.ts
import { Location } from "./location.ts";
import { Range } from "./range.ts";
import { Module } from "./syscall.ts";
import * as typescript from "./typescript.ts";
import ts from "typescript";

export type Request = {
	module: Module;
};

export type Response = {
	symbols: Array<Symbol> | null;
};

export type Symbol = {
	name: string;
	detail: string | null;
	kind: Kind;
	tags: Array<Tag>;
	range: Range;
	selectionRange: Range;
	children: Array<Symbol> | null;
};

export type Kind =
	| "unknown"
	| "file"
	| "module"
	| "namespace"
	| "package"
	| "class"
	| "method"
	| "property"
	| "field"
	| "constructor"
	| "enum"
	| "interface"
	| "function"
	| "variable"
	| "constant"
	| "string"
	| "number"
	| "boolean"
	| "array"
	| "object"
	| "key"
	| "null"
	| "enumMember"
	| "event"
	| "operator"
	| "typeParameter";

export type Tag = "deprecated";

export let handle = (request: Request): Response => {
	// Get the source file and position.
	let sourceFile = typescript.host.getSourceFile(
		typescript.fileNameFromModule(request.module),
		ts.ScriptTarget.ESNext,
	);

	if (sourceFile === undefined) {
		throw new Error();
	}

	let symbols = [];

	// Get the navigation tree for this file.
	let navigationTree = typescript.languageService.getNavigationTree(sourceFile);

	// Get the symbols by walking the navigation tree.
	let symbols = [walk(navigationTree)];

	return { symbols }
};

// https://github.com/microsoft/TypeScript/blob/59d3a381807bb4247a36a24be7e41553ebe6d8b5/src/services/types.ts#L826
export let walk = (file: string, tree: typescript.NavigationTree): Symbol => {
	let name = tree.text;

	// Find the range of this symbol and its selectionRange.
	let span = tree
		.span
		.reduce((acc, span) => {
			acc.start = Math.min(acc.start, span.start);
			acc.end = Math.max(acc.end, span.end);
		});
	let range = spanToRange(file, span);
	let selectionRange = spanToRange(file, tree.nameSpan ?? span);

	// Parse the symbol kind from the nav tree.
	let kind = getKind(tree.kind);

	// Collect the nested children.
	let children = tree.children?.map(walk);

	return {
		name,
		kind,
		tags: [], // TODO: deprecation tags.
		detail: null, // TODO: symbol details.
		range,
		selectionRange,
		children,
	};
};

// https://github.com/microsoft/TypeScript/blob/59d3a381807bb4247a36a24be7e41553ebe6d8b5/src/services/types.ts#L1559
let getKind = (tsKind:string): Kind => {
	switch(tsKind) {
		case "script": {
			kind = "file";
			break;
		}
		case "module": {
			kind = "module";
			break;
		}
		case "class": {
			kind = "class";
			break;
		}
		case "local class": {
			kind = "class";
			break;
		}
		case "interface": {
			kind = "interface";
			break;
		}
		case "type": {
			kind = "class";
			break;
		}
		case "enum": {
			kind = "class";
			break;
		}
		case "var": {
			kind = "variable";
			break;
		}
		case "local var": {
			kind = "variable";
			break;
		}
		case "function": {
			kind = "function";
			break;
		}
		case "local function": {
			kind = "function";
			break;
		}
		case "method": {
			kind = "method";
			break;
		}
		case "getter": {
			kind = "method";
			break;
		}
		case "setter": {
			kind = "method";
			break;
		}
		case "property": {
			kind = "property";
			break;
		}
		case "accessor": {
			kind = "property";
			break;
		}
		case "constructor": {
			kind = "constructor";
			break;
		}
		case "parameter": {
			kind = "variable";
			break;
		}
		case "type parameter": {
			kind = "typeParameter";
			break;
		}
		case "external module": {
			kind = "module";
			break;
		}
		default: {
			kind = "unknown";
			break;
		}
	}
}

File: src/language/server/symbols.rs
use super::Server;
use crate::{
	error::Result,
	language::{
		self,
		service::symbols::{Kind, Symbol},
	},
	module::Module,
};
use lsp_types as lsp;

impl Server {
	pub async fn symbols(
		&self,
		params: lsp::DocumentSymbolParams,
	) -> Result<Option<lsp::DocumentSymbolResponse>> {
		// Get the module.
		let module = Module::from_lsp(&self.server, params.text_document.uri).await?;

		// Get the document symbols.
		let symbols = module.symbols(&self.server).await?;
		let Some(symbols) = symbols else {
			return Ok(None);
		};

		// Convert the symbols.
		let symbols = symbols.into_iter().map(collect_symbol_tree).collect();

		Ok(Some(lsp::DocumentSymbolResponse::Nested(symbols)))
	}
}

fn collect_symbol_tree(symbol: Symbol) -> lsp::DocumentSymbol {
	let Symbol {
		name,
		detail,
		kind,
		tags,
		range,
		selection_range,
		children,
	} = symbol;

	let kind = match kind {
		Kind::File => lsp::SymbolKind::FILE,
		Kind::Module => lsp::SymbolKind::MODULE,
		Kind::Namespace => lsp::SymbolKind::NAMESPACE,
		Kind::Package => lsp::SymbolKind::PACKAGE,
		Kind::Class => lsp::SymbolKind::CLASS,
		Kind::Method => lsp::SymbolKind::METHOD,
		Kind::Property => lsp::SymbolKind::PROPERTY,
		Kind::Field => lsp::SymbolKind::FIELD,
		Kind::Constructor => lsp::SymbolKind::CONSTRUCTOR,
		Kind::Enum => lsp::SymbolKind::ENUM,
		Kind::Interface => lsp::SymbolKind::INTERFACE,
		Kind::Function => lsp::SymbolKind::FUNCTION,
		Kind::Variable => lsp::SymbolKind::VARIABLE,
		Kind::Constant => lsp::SymbolKind::CONSTANT,
		Kind::String => lsp::SymbolKind::STRING,
		Kind::Number => lsp::SymbolKind::NUMBER,
		Kind::Boolean => lsp::SymbolKind::BOOLEAN,
		Kind::Array => lsp::SymbolKind::ARRAY,
		Kind::Object => lsp::SymbolKind::OBJECT,
		Kind::Key => lsp::SymbolKind::KEY,
		Kind::Null => lsp::SymbolKind::NULL,
		Kind::EnumMember => lsp::SymbolKind::ENUM_MEMBER,
		Kind::Event => lsp::SymbolKind::EVENT,
		Kind::Operator => lsp::SymbolKind::OPERATOR,
		Kind::TypeParameter => lsp::SymbolKind::TYPE_PARAMETER,
	};

	let tags = tags
		.into_iter()
		.map(|tag| match tag {
			language::service::symbols::Tag::Deprecated => lsp::SymbolTag::DEPRECATED,
		})
		.collect();

	let children = children.map(|children| children.into_iter().map(collect_symbol_tree).collect());

	let range = lsp::Range {
		start: lsp::Position {
			line: range.start.line,
			character: range.end.character,
		},
		end: lsp::Position {
			line: range.end.line,
			character: range.end.character,
		},
	};

	let selection_range = lsp::Range {
		start: lsp::Position {
			line: selection_range.start.line,
			character: selection_range.end.character,
		},
		end: lsp::Position {
			line: selection_range.end.line,
			character: selection_range.end.character,
		},
	};

	#[allow(deprecated)]
	lsp::DocumentSymbol {
		name,
		detail,
		kind,
		tags: Some(tags),
		range,
		selection_range,
		children,
		deprecated: None,
	}
}
